### PR TITLE
applets/places-indicator: fixes and new features

### LINKED ISSUE #673: Places indicator: Open the file manager while trying to mount an encrypted device
**Operating System**: Solus 1.2.1
**Budgie version**: Budgie 10.2.8

**The issue encountered**
When trying to open an encrypted device on Places indicator applet, it displays an error: "G_MOUNT_OPERATION_HANDLED but got 2"
In these cases it is better to open the file manager and being entrusted it to display a dialog to enter a password or add a layer of integration with budgie.

**Steps to reproduce the issue**
- Select an encrypted partition/flash drive/SD/external hdd
- Try to mount/open with Places indicator applet

### PR DESCRIPTION
**Pull request by cybre**
This PR contains a lot of changes to different components (sorry, it really should have been multiple commits but shit happened)
Some of the changes:
 - The applet can now unlock encrypted volumes via the new mount helper (a password entry appears when trying to mount an encrypted volume) fixes #673
 - S/FTP and SMB shares are now displayed in the applet
 - Clicking on S/FTP bookmarks now opens them in the default file manager for mounting
 - Structural and general code enhancements
 - Style and display improvements
 - The places section now contracts when the applet popover is closed

### CODE REVIEW

File: src/panel/applets/places-indicator/MountHelper.vala [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    private void handle_aborted()
+    {
+        message("Operation aborted");
+        this.reply(GLib.MountOperationResult.HANDLED);
+    }
+
+    private void handle_question()
+    {
+        message("Questions with multiple choices are not supported at the moment");
+        this.reply(GLib.MountOperationResult.HANDLED);
```
Comment (ikeydoherty) [inline src/panel/applets/places-indicator/MountHelper.vala]: So then how is it handled?

File: src/panel/applets/places-indicator/MountHelper.vala [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+        password_entry.set_input_purpose(Gtk.InputPurpose.PASSWORD);
+        password_entry.set_visibility(false);
+        unlock_box.pack_start(password_entry, true, true, 0);
+        Gtk.Button unlock_button = new Gtk.Button.from_icon_name("changes-allow-symbolic", Gtk.IconSize.MENU);
+        unlock_button.set_sensitive(false);
+        unlock_box.pack_end(unlock_button, false, false, 0);
+
+        revealer.show_all();
+
+        password_entry.changed.connect(()=> {
```
Comment (ikeydoherty) [inline src/panel/applets/places-indicator/MountHelper.vala]: lambdas are hard to debug

File: src/panel/applets/places-indicator/PlacesIndicatorWindow.vala [hunk 3]
```diff
@@ -345,44 +392,44 @@ public class PlacesIndicatorWindow : Gtk.Popover {
...
     }
 
     /*
      * Adds a mount to the view
      */
-    private void add_mount(GLib.Mount mount, string? class)
+    private void add_mount(GLib.Mount mount, string? mount_class)
     {
-        if ((class == "network" && !show_networks) || (class == "device" && !show_drives)) {
+        if ((!mount.can_unmount() && !mount.can_eject()) || mount.is_shadowed()) {
```
Comment (ikeydoherty) [inline src/panel/applets/places-indicator/PlacesIndicatorWindow.vala]: This is hard to interpret.

File: src/panel/applets/places-indicator/PlacesIndicatorWindow.vala [hunk 4]
```diff
@@ -392,21 +439,20 @@ public class PlacesIndicatorWindow : Gtk.Popover {
...
-        }
+        string new_path = path.split(" ")[0];
+        GLib.File file = GLib.File.new_for_uri(new_path);
 
-        GLib.File file = GLib.File.new_for_uri(path);
         PlaceItem place_item = new PlaceItem(file, "place");
         places_list.add(unescaped_path);
         places_section.add_item(place_item);
+
+        place_item.send_message.connect((message_content)=> {
```
Comment (ikeydoherty) [inline src/panel/applets/places-indicator/PlacesIndicatorWindow.vala]: lambdas..

File: src/panel/applets/places-indicator/VolumeItem.vala [hunk 5]
```diff
@@ -35,10 +37,25 @@ public class VolumeItem : ListItem
...
+        operation = new MountHelper();
+
+        operation.send_message.connect((message)=> {
+            send_message(message);
+        });
+
+        operation.password_asked.connect(()=> {
+            if (unlock_revealer == null) {
+                unlock_revealer = operation.get_encryption_form();
+                base.pack_start(unlock_revealer, true, true, 0);
```
Comment (ikeydoherty) [inline src/panel/applets/places-indicator/VolumeItem.vala]: Why base and not this? Also.. lambdas..

File: src/panel/applets/places-indicator/VolumeItem.vala [hunk 6]
```diff
@@ -55,28 +72,63 @@ public class VolumeItem : ListItem
...
+                unlock_revealer.set_reveal_child(!unlock_revealer.get_child_revealed());
+                operation.get_password_entry().grab_focus();
+            }
         });
     }
+
+    private void do_mount(GLib.Volume volume)
+    {
+        spin.start();
+        volume.mount.begin(GLib.MountMountFlags.NONE, operation, null, (obj, res)=>{
```
Comment (ikeydoherty) [inline src/panel/applets/places-indicator/VolumeItem.vala]: This really should not be a lambda, the size of it is inappropriate. Should be a real callback that can be debugged

**Comment by cybre** [comment]
All problematic lambdas have been replaced with proper callbacks.

**Comment by ikeydoherty** [comment]
Sorry, brianfart. Done!

**Pull request merged**